File: AHYEON/17./pro_가장먼노드.swift
import Foundation

func solution(_ n:Int, _ edge:[[Int]]) -> Int {
    var graph: [Int: [Int]] = [:]

    func connect(key: Int, value: Int) {
        if graph.keys.contains(key) {
            graph[key]?.append(value)
        } else {
            graph[key] = [value]
        }
    }

    for e in edge {
        connect(key: e[0], value: e[1])
        connect(key: e[1], value: e[0])
    }

    var Q: [(Int, Int)] = []
    var answer: [(Int, Int)] = []

    var visited: [Bool] = Array(repeating: false, count: graph.keys.count)

    func bfs(key: Int, depth: Int) {
        for node in graph[key]! {
            if visited[node-1] {
                continue
            }
            visited[node-1] = true
            answer.append((node, depth))
            Q.append((node, depth))
            }
        }
        Q.append((1,0))
        visited[0] = true
        while !Q.isEmpty {
            let node = Q.removeFirst()
            bfs(key: node.0, depth: node.1+1)
        }
        let maxNum = answer.max {$0.1 < $1.1}.map {$0.1}
        var count = 0
        for node in answer {
            if maxNum == node.1 {
            count += 1
            }
        }
        return count

}

File: 염성훈/17_Programmers/pro_가장먼노드.py
from collections import deque

def bfs(visited, graph):
    q = deque()
    q.append(1)
    visited[1] = 1

    while q:
        node = q.popleft()
        for w in graph[node]:
            if not visited[w]:
                visited[w] += visited[node] + 1
                q.append(w)


def solution(n, edge):
    answer = 0
    graph = [[] for _ in range(n + 1)]
    for line in edge:
        graph[line[0]].append(line[1])
        graph[line[1]].append(line[0])

    visited = [0] * (n + 1)
    bfs(visited, graph)

    for i in visited:
        if i == max(visited):
            answer += 1

    return answer

n = 6
vertex = [[3, 6], [4, 3], [3, 2], [1, 3], [1, 2], [2, 4], [5, 2]]
solution(n, vertex)